src/backend/distributed/utils/node_metadata.c

```diff
@@ -620,27 +683,43 @@ AddNodeMetadata(char *nodeName, int32 nodePort, int32 groupId, char *nodeRack,
 		}
 	}
 
+	/* if nodeRole hasn't been added yet there's a constraint for one-node-per-group */
+	if (nodeRole != InvalidOid)
+	{
+		if (nodeRole == PrimaryNodeRoleId())
+		{
+			WorkerNode *existingPrimaryNode = PrimaryNodeForGroup(groupId, NULL);
+
+			if (existingPrimaryNode != NULL)
+			{
+				ereport(ERROR, (errmsg("group %d already has a primary node", groupId)));
+			}
+		}
+	}
+
 	/* generate the new node id from the sequence */
 	nextNodeIdInt = GetNextNodeId();
 
 	InsertNodeRow(nextNodeIdInt, nodeName, nodePort, groupId, nodeRack, hasMetadata,
-				  isActive);
+				  isActive, nodeRole);
 
 	workerNode = FindWorkerNode(nodeName, nodePort);
 
-	/* send the delete command all nodes with metadata */
+	/* send the delete command to all primary nodes with metadata */
 	nodeDeleteCommand = NodeDeleteCommand(workerNode->nodeId);
 	SendCommandToWorkers(WORKERS_WITH_METADATA, nodeDeleteCommand);
 
 	/* finally prepare the insert command and send it to all primary nodes */
-	workerNodeList = list_make1(workerNode);
-	nodeInsertCommand = NodeListInsertCommand(workerNodeList);
-	SendCommandToWorkers(WORKERS_WITH_METADATA, nodeInsertCommand);
+	primariesWithMetadata = CountPrimariesWithMetadata();
+	if (primariesWithMetadata != 0)
+	{
+		List *workerNodeList = list_make1(workerNode);
+		char *nodeInsertCommand = NodeListInsertCommand(workerNodeList);
+		SendCommandToWorkers(WORKERS_WITH_METADATA, nodeInsertCommand);
+	}
 
-	heap_close(pgDistNode, AccessExclusiveLock);
+	heap_close(pgDistNode, NoLock);
```

Comment:
Should be careful with this. The function takes an access exclusive lock on pg_dist_node and now keeps it until commit, which blocks all queries. Maybe take an ExclusiveLock.